src/cargo/ops/cargo_compile.rs

```diff
@@ -1455,3 +1462,108 @@ fn opt_patterns_and_names(
     }
     Ok((opt_patterns, opt_names))
 }
+
+/// Removes duplicate CompileMode::Doc units that would cause problems with
+/// filename collisions.
+///
+/// Rustdoc only separates units by crate name in the file directory
+/// structure. If any two units with the same crate name exist, this would
+/// cause a filename collision, causing different rustdoc invocations to stomp
+/// on one another's files.
+///
+/// Unfortunately this does not remove all duplicates, as some of them are
+/// either user error, or difficult to remove. Cases that I can think of:
+///
+/// - Same target name in different packages. See the `collision_doc` test.
+/// - Different sources. See `collision_doc_sources` test.
+///
+/// Ideally this would not be necessary.
+fn remove_duplicate_doc(build_config: &BuildConfig, unit_graph: &mut UnitGraph) {
```

Comment:
FWIW long-term I don't think that this function should exist. Ideally rustdoc would have a feature that we could specify the directory/lookup name and we could pass that in to avoid all duplication issues.

Reply:
I agree. However, it is not clear to me how that should work.  Because the directories are exposed as URLs, I think it would be unpleasant for those to have Cargo-style hashes in them.  There could be some pattern like $NAME-$VERSION and then have extra disambiguators when necessary like $NAME-$VERSION-bin-foo.  But there are some collisions that are harder (like when they only differ in the features).

What do you think about having a hash in the URL?  

This feels like a bit of a mess, and I'm uncertain how to climb out of it.

Reply:
At least for `cargo doc` locally I feel like the hash doesn't matter too much since you'll just be clicking on links anyway (assuming we generate some well-known landing page which we don't already do today). For everything else though (like doc.rust-lang.org or docs.rs) I wouldn't want Cargo-style hashes in URLs. 

This function is likely gonna be here for a long long time which is by no means the end of the world. I think it'll take some more thinking to figure out what it might look like to not have this function.